graphql.go

```diff
@@ -73,12 +76,143 @@ func (c *Client) logf(format string, args ...interface{}) {
 	c.Log(fmt.Sprintf(format, args...))
 }
 
+// RetryConfig defines possible fields that client can supply for their retry strategies
+type RetryConfig struct {
+	// Optional - Max number of times client should retry
+	MaxTries int `json:"maxTries"`
+	// Required - Time interval to wait before trying attempt sending a request again
+	Interval float64 `json:"interval"`
+	// Required - Defines a policy to be used for retry
+	Policy PolicyType `json:"policy"`
+	// Optional - The max interval of time to wait before retrying
+	MaxInterval float64 `json:"maxInterval"`
+	// Optional - A mapping of statuses that client should retry.
+	// If not specifed, we will use default retry behavior on certain statuses
+	RetryStatus map[int]bool `json:"statusToRetry"`
+}
+
+// PolicyType defines a type of different possible Policies to be applied towards retrying
+type PolicyType string
+
+const (
+	// ExponentialBackoff - the interval is doubled after every try until hitting MaxInterval or MaxTries
+	ExponentialBackoff PolicyType = "exponential_backoff"
+	// Linear - the interval stays the same every try until hitting MaxTries
+	Linear PolicyType = "linear"
+)
+
+var (
+	defaultLinearRetryConfig = RetryConfig{
+		MaxTries: 5,
+		Interval: 2,
+		Policy:   Linear,
+	}
+
+	defaultExponentialRetryConfig = RetryConfig{
+		MaxTries:    4,
+		Interval:    1,
+		Policy:      ExponentialBackoff,
+		MaxInterval: 16,
+	}
+)
+
+// Wrapper method to send request while optionally applying retry policy
+func (c *Client) sendRequest(req *http.Request) (*http.Response, error) {
+	var (
+		resp *http.Response
+		err  error
+	)
+
+	retryConfig := c.retryConfig
+	// Client did not specify retry config
+	if retryConfig.Policy == "" {
+		resp, err = c.httpClient.Do(req)
+		if err != nil {
+			return resp, fmt.Errorf("Error getting response: %v", err)
+		}
+		return resp, nil
+	}
+
+	for tryCount := 0; tryCount < retryConfig.MaxTries; tryCount++ {
+		resp, err = c.httpClient.Do(req)
+		if err == nil && !retryConfig.shouldRetry(resp.StatusCode) {
+			return resp, nil
+		}
+		c.Log("Will retry after interval expires")
+
+		// Wait for interval
+		c.logf("Waiting for interval(%f) to expire...", retryConfig.Interval)
+		timer := time.NewTimer(time.Duration(retryConfig.Interval) * time.Second)
+
+		ctx := req.Context()
+
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("Context finished unexpectedly")
+
+		case <-timer.C:
+			// Increase interval
+			retryConfig.increaseInterval()
+			c.logf("New interval: %f", retryConfig.Interval)
+		}
+
+	}
+
+	return nil, fmt.Errorf("Error getting response with retry: %s", err)
+}
+
+// Increase interval for exponential backoff policy until hitting MaxInterval
+func (config *RetryConfig) increaseInterval() {
+	if config.Policy == ExponentialBackoff && config.Interval < config.MaxInterval {
+		config.Interval = math.Min(config.Interval*2, config.MaxInterval)
```

Comment:
I think there might be a bug here.  You are updating config.Interval directly and since config is a pointer, the next time Interval is used it retains the value from the last time it was updated instead of starting from the original value.   It is better to have a generic getNextInterval() function that returns an int value for next interval to use and assign that to a local variable rather than overriding the original config.Interval.  Since you're already merged please create a new PR to fix this bug.

Reply:
@tnguyen-veri I initially thought the same thing, but he is making a copy of the config value on line 126 
https://github.com/veritone/graphql/pull/1/files#diff-599ffce354627452a71923419f5f7501R126

Reply:
Ah yes you are correct. Nvm then.